prototypes/palloc/pool_alloc.c

```diff
@@ -0,0 +1,212 @@
+    return found_pi -> pi_data;
+} 
+
+void pool_init(pool_t *pool, size_t size, 
+        void (*constructor)(void *, size_t), void (*destructor)(void *, size_t)) {
+    /*
+    init object and give it 1-page cache;
+    */
+    DEBUG_PRINT("Entering pool_init\n");
+    size = align_to_binary_pow(size, 3); //Align to 64-bit (8 byte) word
+    LIST_INIT(&(pool -> pp_empty_slabs));
+    LIST_INIT(&(pool -> pp_full_slabs));
+    LIST_INIT(&(pool -> pp_part_slabs));
+    pool_slab_t *first_slab = create_slab(size, constructor);
+    LIST_INSERT_HEAD(&(pool -> pp_empty_slabs), first_slab, ph_slablist);
+    pool -> pp_constructor = constructor;
+    pool -> pp_destructor = destructor;
+    pool -> pp_itemsize = size;
+    pool -> pp_nslabs = 1;
+    pool -> pp_nitems = first_slab -> ph_ntotal;
+    DEBUG_PRINT("Initialized new pool at 0x%lx\n", (uint64_t) pool);
+}
+
+
+void pool_destroy(pool_t* pool) {
+    /*
+    take some objects from the cache;
+    destroy the objects;
+    free the underlying memory;
+    */
+
+    DEBUG_PRINT("Entering pool_destroy\n");
+
+    uint64_t *tmp = (uint64_t*) pool;
+
+    for (size_t i = 0; i < sizeof(pool_t)/8; i++) {
+        if (tmp[i] == 0xdeadbeefdeadbeef) {
+            panic("double free at pool 0x%lx\n", (uint64_t) pool);
+        }
+    }
+
+    DEBUG_PRINT("Destroying empty slabs\n");
+    pool_slab_t *it = LIST_FIRST(&(pool -> pp_empty_slabs));
+    while(it) {
+        pool_slab_t *next = LIST_NEXT(it, ph_slablist);
+        LIST_REMOVE(it, ph_slablist);
+        destroy_slab(it, pool);
+        it = next;
+    }
+
+    DEBUG_PRINT("Destroying partially filled slabs\n");
+    it = LIST_FIRST(&(pool -> pp_part_slabs));
+    while(it) {
+        pool_slab_t *next = LIST_NEXT(it, ph_slablist);
+        LIST_REMOVE(it, ph_slablist);
+        destroy_slab(it, pool);
+        it = next;
+    }
+
+    DEBUG_PRINT("Destroying full slabs\n");
+    it = LIST_FIRST(&(pool -> pp_full_slabs));
+    while(it) {
+        pool_slab_t *next = LIST_NEXT(it, ph_slablist);
+        LIST_REMOVE(it, ph_slablist);
+        destroy_slab(it, pool);
+        it = next;
+    }
+
+    for (size_t i = 0; i < sizeof(pool_t)/8; i++) {
+        tmp[i] = 0xdeadbeefdeadbeef;
+    }
+    DEBUG_PRINT("Destroyed pool at 0x%lx\n", (uint64_t) pool);
+}
+
+void *pool_alloc(pool_t* pool, __attribute__((unused)) uint32_t flags) { //TODO: implement different flags
+    /*
+    if (there’s an object in the cache)
+        take it (no construction required);
+    else {
+        allocate memory;
+        construct the object;
+    }
+    */
+    DEBUG_PRINT("Entering pool_alloc\n");
+
+    pool_slab_t *new_slab;
+    if (pool -> pp_nitems) {
+        if (LIST_EMPTY(&(pool -> pp_part_slabs))) {
+            new_slab = LIST_FIRST(&(pool -> pp_empty_slabs));
+            LIST_REMOVE(new_slab, ph_slablist);
+        } else {
+            new_slab = LIST_FIRST(&(pool -> pp_part_slabs));
+            LIST_REMOVE(new_slab, ph_slablist);
+        }
```

Comment:
These two `if` cases have almost identical content. Maybe it would be simpler to write:

```c
pool_slab_list_t* slab_list_to_use = pool->pp_nitems ? &pool->pp_empty_slabs : &pool->pp_part_slabs;
new_slab = LIST_FIRST(slab_list_to_use);
LIST_REMOVE(new_slab, ph_slablist);
```
or maybe even:
```c
new_slab = LIST_FIRST(pool->pp_nitems ? &pool->pp_empty_slabs : &pool->pp_part_slabs);
LIST_REMOVE(new_slab, ph_slablist);
```
Similar transformation can be used to simplify many of your `if` statements - but I am not sure if the above suggestion fits our coding style.